WalletWasabi/Crypto/ZeroKnowledge/LinearRelation/Equation.cs

```diff
@@ -0,0 +1,61 @@
+using NBitcoin.Secp256k1;
+using System.Linq;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge.LinearRelation
+{
+	// Each proof of a linear relation consists of multiple knowledge of
+	// representation equations, all sharing a single witness comprised of several
+	// secrets.
+	//
+	// Note that some of the generators can be the point at infinity, when a term
+	// in the witness does not play a part in the representation of a specific
+	// point.
+	public class Equation
+	{
+		public Equation(GroupElement publicPoint, GroupElementVector generators)
+		{
+			CryptoGuard.NotInfinity(nameof(generators), generators);
+			PublicPoint = CryptoGuard.NotInfinity(nameof(publicPoint), publicPoint);
+			Generators = generators;
+		}
+
+		// Knowledge of representation asserts
+		//     P = x_1*G_1 + x_2*G_2 + ...
+		// so we need a single public input and several generators
+		public GroupElement PublicPoint { get; }
+		public GroupElementVector Generators { get; }
+
+		// Evaluate the verification equation corresponding to the one in the statement
+		internal bool Verify(GroupElement publicNonce, Scalar challenge, ScalarVector responses)
+		{
+			// the verification equation (for 1 generator case) is:
+			//   sG =? R + eP
+			// where:
+			//   - R = kG is the public nonce, k is the secret nonce
+			//   - P = xG is the public input, x is the secret
+			//   - e is the challenge
+			//   - s is the response
+			return (publicNonce + challenge * PublicPoint) == responses * Generators;
```

Comment:
nit: flip so is in same order as above equation `sG =? R + eP`
```
return responses * Generators == (publicNonce + challenge * PublicPoint);
```

Reply:
Also, you need to check that `responses != 0` in case it is not the point of infinity. Otherwise any signature can be made valid by having a `0` `s` value

Reply:
i'm not sure that's true after writing a test for this.

for the RHS to also be the point at infinity, the nonce has to be `challenge.Negate() * PublicPoint`, but in both the interactive setting and with Fiat-Shamir the prover does not know the challenge yet when comitting to the nonce.

however, i don't see the harm in adding this check, assuming the secret nonce is sampled uniformly from the scalar field, `publicNonce` should be uniform in the group, which means the probability of the RHS being the point at infinity should be negligible, and consequentially so should the probability of one of the response components being 0.

@seresistvanandras

Reply:
As Yuval points it out, I also don't see the necessity of adding a guard/check here. Even if `response == 0`, the special soundness property of the sigma protocol is maintained in both the interactive and non-interactive versions of the proof system. In the non-interactive case, if one wants to prove a false statement, then she needs to essentially find a collision in a cryptographically secure hash function.

For a given `response` value `s`, there is a single challenge value `e`, which will make the proof verified. Therefore, malicious prover essentially needs to find the preimage `x` such that `H(x)=e`. No value `e` has a distinguished role when one tries to invert a cryptographically secure hash function. For each fixed `e` this problem will be computationally infeasible. Therefore I would not consider adding such a guard.

Reply:
@benthecarman thanks. We switched the equality sides to have it in the same order than the doc.